Show marker context menu in the timeline

gribnoysup opened the pull request:
This is a follow-up to the #2269. Most of the implementation details that are in this PR were discussed there.

There are three main changes in this PR:

- Move `rightClickedCallNode` state out of per-thread one
- Move `rightClickedMarker` state out of per-thread one
- Add marker context menu triggers and related actions to the timeline marker components

Most of the tests didn't require any change as the code changes were pure refactor. I haven't added any new tests for the context menu in the timeline but planning to do so, just wanted to open the PR as it is mostly ready for review and I would appreciate some feedback.

Closes #2189

TODO:

- [x] Add tests for the marker context menu in the timeline

gribnoysup commented on src/components/shared/CallNodeContextMenu.js:
```diff
@@ -502,16 +515,32 @@ class CallNodeContextMenu extends PureComponent<Props> {
+      threadIndex !== null && getThreadSelectors(threadIndex);
+
+    return {
+      rightClickedCallNode:
```
This was done a little bit differently in the original PR, but the more I thought about it the more it made sense to completely switch call node context menu to use the same thread as the right-clicked call node has instead of mixing selected thread with right clicked call node (even though currently they probably always be the same)

gribnoysup commented on src/reducers/profile-view.js:
```diff
@@ -499,6 +447,96 @@ const rightClickedTrack: Reducer<TrackReference | null> = (
+      const functionUpdate = functionsUpdatePerThread[state.threadIndex];
+
+      if (!functionUpdate) {
+        console.error(
```
As mentioned in https://github.com/firefox-devtools/profiler/pull/2269#discussion_r344297102 I added a warning message and a state reset for this reducer when there is no `functionUpdate` for the thread found, but now I'm wondering should something like this also be applied to [other](https://github.com/firefox-devtools/profiler/blob/c75517fad718995ff3c87a53291a2fa32cf33140/src/reducers/profile-view.js#L45-L47) [places](https://github.com/firefox-devtools/profiler/blob/c75517fad718995ff3c87a53291a2fa32cf33140/src/reducers/profile-view.js#L143-L145) where reducer doesn't have a `functionUpdate` for the thread? Currently those uther reducers just keep the state intact if that happens

julienw replied:
I think I was wrong before: it's totally normal that we can't find a thread index.
You can look at the class `ColascedFunctionsUpdateDispatcher` in `src/actions/receive-profile`: this happens during load time, when we symbolicate the profile, and we find that some functions need to be merged (when 2 symbols A and symbols B resolve to the same function). This doesn't necessarily happen in all threads. (I don't know the exact situations when this happens or doesn't happen).

julienw replied:
You can also look at ` finishSymbolicationForLib` in `src/profile-logic/symbolication`.

gribnoysup replied:
Ahh, yes, makes sense! I'll change it then to just return the current state if there is no `functionUpdate`

julienw commented on src/selectors/right-clicked-marker.js:
```diff
@@ -0,0 +1,43 @@
+  getProfileViewOptions,
+  ({ rightClickedMarker }) =>
+    rightClickedMarker ? rightClickedMarker.markerIndex : null
+);
```
Of course maybe you'll just remove these functions if you follow the other comments, but I wanted to point that out :)

gribnoysup replied:
Thanks for those comments! They are helpful, educating and definitely help me to make this PR better. And yes, while refactoring the selectors as suggested in the other comments I found no use for those simple selectors and removed them 😄

gribnoysup commented on src/components/shared/CallNodeContextMenu.js:
```diff
@@ -502,16 +515,32 @@ class CallNodeContextMenu extends PureComponent<Props> {
+              callNodePath,
+              callNodeIndex,
+            }
+          : null,
```
Yep, this makes sense, I'll refactor the selectors 👍

julienw commented on src/selectors/right-clicked-call-node.js:
```diff
@@ -0,0 +1,89 @@
+
+    return null;
+  }
+);
```
`createSelector` works this way:
```
createSelector(
  selector1,
  selector2,
  selector3,
  (result1, result2, result3) => result
);
```
`result1` is the result of `selector1`, etc.
`selector1` is a function that takes the full `state` and returns something, but it's usually just another already existing selector.

Hope this helps!

gribnoysup replied:
I defined selectors for `thread,` `callNodeIndex` and `callNodeInfo` right in the `createSelector` as they are not used by anything else in the code. Do you think I should move them to separate methods?

gribnoysup commented on src/test/components/TimelineMarkers.test.js:
```diff
@@ -148,4 +229,95 @@ describe('TimelineMarkers', function() {
     delete window.devicePixelRatio;
   });
+
+  describe('displays context menus', () => {
```
Those new tests that I added are the same as for the context menu in MarkerChart. This required moving a bunch of stuff around though, this is why diff is so big 😓

gribnoysup commented:
Hey @julienw! I was wondering if you think that there are more changes required in this PR before it can be merged?

julienw commented:
Ah, I'll look at it soon, sorry I missed the pushes!

julienw commented on src/types/state.js:
```diff
@@ -37,10 +37,18 @@ export type Reducer<T> = (T | void, Action) => T;
-  +rightClickedMarker: MarkerIndex | null,
+|};
+
+export type RightClickedCallNodePath = {|
```
I think the type shouldn't be "...CallNodePath" because it's not _just_ a call node path. A call node path is something very specific in our code, this is an array of function indexes, that uniquely identifies a specific call node. So this is confusing that we see `RightClickedCallNodePath` but it's really an object containing things.
I'd suggest `RightClickedCallNode`.
The same remark goes to the reducer/state property name as well.

julienw commented on src/selectors/right-clicked-call-node.js:
```diff
@@ -0,0 +1,89 @@
+    }
+
+    return null;
+  },
```
The same remark goes for the marker selector below.

julienw commented on src/components/shared/MarkerContextMenu.js:
```diff
@@ -37,8 +38,7 @@ import { getMarkerFullDescription } from '../../profile-logic/marker-data';
   +committedRange: StartEndRange,
-  +selectedMarker: Marker | null,
-  +thread: Thread,
+  +rightClickedMarker: RightClickedMarkerInfo | null,
```
nit: maybe call this property `rightClickedMarkerInfo`. In this project we try to have very descriptive names, even if this makes them long.

julienw commented on src/components/shared/CallNodeContextMenu.js:
```diff
@@ -31,22 +30,13 @@ import {
-  +thread: Thread,
-  +threadIndex: ThreadIndex,
-  +callNodeInfo: CallNodeInfo,
+  +rightClickedCallNode: RightClickedCallNodeInfo | null,
```
same remark as for the marker context menu: maybe call this `rightClickedCallNodeInfo` so that in the code we better know the type of the variables we're dealing with.

julienw commented on src/components/calltree/CallTree.js:
```diff
@@ -240,31 +241,35 @@ class CallTreeComponent extends PureComponent<Props> {
+      ),
+      rightClickedCallNodeIndex: callNodeInfo
+        ? callNodeInfo.callNodeIndex
+        : null,
```
for more correctness, maybe we should pass `null` if the selected threadIndex is different than the threadIndex in the callNodeInfo. What do you think?

Of course this is the same in other similar locations.
Because the code will be the same, maybe we should extract a simple function:
```js
function getRightClickedCallNodeIndexForThread(state, threadIndex) {
   const callNodeInfo = getRightClickedCallNodeInfo(state);
  if (!callNodeInfo || callNodeInfo.threadIndex !== threadIndex) {
    return null;
  }
  return callNodeInfo.callNodeIndex;
}
```
I'm thinking that maybe this function could even be in `selectedThreadSelectors`. This way we could still have `selectedThreadSelectors.getRightClickedCallNodeIndex`. (it would be slightly different, that it wouldn't take threadIndex as parameter).

What do you think? Would that work?
This is all to make the code more readable in the end.

gribnoysup replied:
Yeah, good catch, this makes a lot of sense! Having a helper seems to be pretty straigtforward, but moving it to `selectedThreadSelectors` will cause a circular dependency between per-thread selectors and right-clicked selectors in the current implementation. I'll look at this case more closely and see if this can be resolved somehow 🤔

julienw commented on src/components/marker-table/index.js:
```diff
@@ -234,19 +235,20 @@ class MarkerTable extends PureComponent<Props> {
-    zeroAt: getZeroAt(state),
-  }),
+  mapStateToProps: state => {
+    const markerInfo = getRightClickedMarkerInfo(state);
```
The same remark about comparing threadIndex applies here too.

julienw commented:
Oh and also it would be nice to get tests for the new reducers and selectors (tested together). This should probably be in `src/test/store/profile-view.test.js`. Ideally we should have tests for each cases, like the right clicked info's thread index is different than the selected thread index, and things like that.

Thanks again!

gribnoysup commented:
You definitely don't need to apologize and thanks again for all the feedback! Unfortunately, I don't think I'll be able to get to that PR until after the holidays, but I'm still happy to improve it and finish the work. I'll ping you when it's done 👍

gribnoysup commented:
Planning to finally finish this today. Outlining what's left to do so I can keep track easier:

- [x] Rename `RightClickedCallNode` (https://github.com/firefox-devtools/profiler/pull/2305#discussion_r357248940)
- [x] Give more explicit names to props so they represent the type of the prop better (https://github.com/firefox-devtools/profiler/pull/2305#discussion_r357250317)
- [ ] Simplify selectors implementation (https://github.com/firefox-devtools/profiler/pull/2305#discussion_r357214576)
- [ ] Be more strict about passing `rightClickedCallNodeIndex` in `CallTree` and `MarkerTable` components (https://github.com/firefox-devtools/profiler/pull/2305#discussion_r357214576)
- [ ] Add tests for new selectors and reducers (https://github.com/firefox-devtools/profiler/pull/2305#issuecomment-565101664)

gribnoysup commented:
Hey @julienw so I think I managed to address most of the feedback that you provided last time. Some things that I wanted to highlight:

> I'm thinking that maybe this function could even be in `selectedThreadSelectors`. This way we could still have `selectedThreadSelectors.getRightClickedCallNodeIndex`. (it would be slightly different, that it wouldn't take threadIndex as parameter).

I didn't manage to add the newly added `getRightClicked*ForThread` selectors to the selected thread selectors as this was causing a circular dependency error between them and I couldn't figure out a good way to restructure the code to make it work.

> Ideally we should have tests for each cases, including the right clicked info's thread index is different than the selected thread index, and things like that.

I added as much tests as I could, but they are pretty basic. I'm struggling to add more as some action handlers that those new reducers react to require very specific information to be provided if I want to dispatch them. For example if I would like to add a test that right-clicked marker is reset when `CHANGE_INVERT_CALLSTACK` action is dispatched, I'll first need to create this action with all the correct parameters passed to the action creator and I'm just lacking knowledge of most of them to do this right. If the tests that I added are not enough I'll need some help to finish this part.

I tried to make commits focused so they are easy to review and I hope I managed to do so 😊

I also added commit references to this list that I made for myself to keep track of the feedback that you provided and the work that I had to do: https://github.com/firefox-devtools/profiler/pull/2305#issuecomment-575887445

julienw commented:
Just a quick note that I've seen your review request and will look at it shorly. Thanks again!

julienw commented on src/selectors/right-clicked-call-node.js:
```diff
@@ -0,0 +1,76 @@
+|};
+
+export const getRightClickedCallNodeInfo: Selector<RightClickedCallNodeInfo | null> = createSelector(
+  state => state,
```
so, I know I suggested this in my previous comment. Now I'm not so sure about that anymore.
The problem is that this breaks memoization: memoization will only work if `state` is the same.

What problem does that give? Mainly that all the code will be run more than necessary.
Is that a big problem? Depends on the code :-)

We can decompose it:
* getProfileViewOptions => it's memoized elsewhere
* thread selectors, getFilteredThread, getCallNodeInfo => memoized elsewhere
* getCallNodeIndexFromPath => not memoized
* return object => not memoized

The fact that the return object isn't memoized will trigger needless rerenders on the context menu component... except in our specific case we return null very quickly most of the time.

I need to look at it closer before giving a more helpful advice.
One obvious question is: should we move these selectors to selectedThreadSelectors after all? (while keeping the state where you've put it, which was the main problem in the previous code). Then it would go to stack-sample.js for thi one, and markers.js for the marker selector.

Does that matter anyway, given this returns null most of the time?

Would you be open to try this change and see if this works?

gribnoysup replied:
Yeah, sure, I'll try to do that! After thinking about moving this stuff to thread selectors I think it wouldn't be as tricky to refactors the components afterwards as I thiught initially as not that many of them actually need anything but indexes that are already available in the right-clicked reducer 👍

julienw commented on src/test/store/profile-view.test.js:
```diff
@@ -32,6 +32,14 @@ import * as ProfileView from '../../actions/profile-view';
 import { viewProfile } from '../../actions/receive-profile';
```
Thanks for the new tests!

gribnoysup commented:
Thanks for another round of feedback @julienw! I think the new implementation covers all the points you mentioned

mstange commented:
Bug report from @hotsphink (originally filed as [bug 1621709](https://bugzilla.mozilla.org/show_bug.cgi?id=1621709)):
> I'm not sure exactly how to describe this, but the last couple of times I've used the profiler I've gotten into a stuck state. I click and drag across the tracks in order to zoom into a region, but it won't "let go" when I release the mouse button -- if I'm dragging from left to right, the right boundary keeps moving after I release the mouse button.
>
> More precisely: I press down the mouse button at some point in the tracks. I drag to the right, and release. The little icon to select that zoom appears in the center of the zoomed region, as expected. I move the mouse to try to click it, and the right side of the boundary moves with the mouse as if I hadn't released.
> 
> Re-clicking or re-dragging doesn't help. If I select a text entry widget and start typing text, the zoom icon reappears until I move the mouse again, at which point it's back to dragging the selection region.
> 
> If I do drag out a new region and release the mouse, it will use the start of the first region and the end of the second, and it will still be hanging onto the the right side when I move the mouse.
> 
> Reloading resets things.
> 
> Aha! It's when I drag out a region and release the mouse button outside of the browser window, or more specifically when the browser window loses focus. 100% reproducible that way.

gribnoysup commented:
Hey @mstange I'm sorry to hear that this PR caused bugs. I would love to fix them but I can't reproduce what's described in the bug report following the provided steps. The code seems to be performing the same for me between master and this branch.

Based on that I don't think I can do anything else here and somebody with more expertise should probably try to handle the issue.

Also as the original issue #2189 was auto-closed due to the initial PR merge it's probably should be re-opened now and I should be unassigned.

julienw commented:
Hey Sergey, don't worry I'll look at it tomorrow when I come back from my
holidays.
I think like you that the problem was preexisting (this is a common problem
with web applications), but I'll double check!

Le dim. 15 mars 2020 à 10:23, Sergey <notifications@github.com> a écrit :

> Hey @mstange <https://github.com/mstange> I'm sorry to hear that this PR
> caused bugs. I would love to fix them but I can't reproduce what's
> described in the bug report following the provided steps. The code seems to
> be performing the same for me between master and this branch.
>
> Based on that I don't think I can do anything else here and somebody with
> more expertise should probably try to handle the issue.
>
> Also as the original issue #2189
> <https://github.com/firefox-devtools/profiler/issues/2189> was
> auto-closed due to the initial PR merge it's probably should be re-opened
> now and I should be unassigned.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/firefox-devtools/profiler/pull/2305#issuecomment-599186345>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADO4H4L2B47YLOZC4SNCS3RHSNBNANCNFSM4JL2T6MA>
> .
>

julienw commented:
I don't know if you followed: I fixed the performance issue in #2467 and relanded your code without any change in #2468. Thanks again!
